### PR TITLE
Notes todo and attempt to fix memory leak

### PR DESCRIPTION
**Pull request by davecramer**
This may be too naive.

### CODE REVIEW

File: connection.c [hunk 1]
```diff
@@ -560,12 +560,24 @@ CC_clear_col_info(ConnectionClass *self, BOOL destroy)
...
 		}
-		self->ntables = 0;
+		/* now go through the array of column info again and remove the entries that have been freed */
+		for (i = 0; i < self->ntables ; )
+		{
+			/* is this entry null */
+			if (coli = self->col_info[i], NULL == coli)
+			{
+				/* move the next entry into this space */
+				coli = self->col_info[i+1];
```
Comment (progmachine) [inline connection.c:576]: Potential infinite loop, if two sibling array entries is `NULL`.
Potential run out of array borders, if `i` is index of last array element. ASAN will kick you in this place =)
To do such algorithm of cleaning `NULL` pointers out, you need two index variables: `idst` for place you write non-null pointer, `isrc` for place you get that non-null pointer, and isrc goes forward faster than `idst`.

**Comment by progmachine** [comment]
This fix for memory leak is incorrect. It broke the meaning of `CC_clear_col_info` function - this function is clearing COL_INFO cache from `ConnectionClass` object, when it is obsolete, or connection is closing. Your fix will leave obsolete cache entries.
To fix this leak correctly, we need to restore refcnt lifetime management. I have strong intent to do this, because these leaks causing fails with my project's unit tests.
For now, i have another problem - many psqlodbc native tests are failing, saying that `testtab1` does not exist... (

File: connection.c [hunk 2]
```diff
@@ -560,12 +560,24 @@ CC_clear_col_info(ConnectionClass *self, BOOL destroy)
 					free_col_info_contents(coli);
 					free(coli);
 					self->col_info[i] = NULL;
+					self->ntables--;
```
Comment (progmachine) [inline connection.c:563]: You do not need to decrement `self->ntables` here, it will leave some entries out of this and next for-loop scopes.

**Comment by davecramer** [comment]
> This fix for memory leak is incorrect. It broke the meaning of `CC_clear_col_info` function - this function is clearing COL_INFO cache from `ConnectionClass` object, when it is obsolete, or connection is closing. Your fix will leave obsolete cache entries. To fix this leak correctly, we need to restore refcnt lifetime management. I have strong intent to do this, because these leaks causing fails with my project's unit tests. For now, i have another problem - many psqlodbc native tests are failing, saying that `testtab1` does not exist... (

Well I need to figure out how to test this.

As for `testtab1` why not use the existing tests ?

**Comment by progmachine** [comment]
> As for `testtab1` why not use the existing tests ?

I am talking about existing tests, located in `test` directory of psqlodbc project:
[regression.diffs.txt](https://github.com/postgresql-interfaces/psqlodbc/files/15307993/regression.diffs.txt)

```
progmachine@progmachine-ws3:~/proj/workspaces/xenlab.workspace/psqlodbc/test$ make installcheck
rm -f regression.diffs
ODBCSYSINI=. ODBCINSTINI=./odbcinst.ini ODBCINI=./odbc.ini ./reset-db < ./sampletables.sql
connected to DSN=psqlodbc_test_dsn;Database=postgres
Dropping and creating database contrib_regression...
connected to DSN=psqlodbc_test_dsn;Database=contrib_regression
Running initialization script...
Done!
prove -e ./runsuite connect stmthandles select update commands multistmt getresult colattribute result-conversions prepare premature params param-conversions parse identity notice arraybinding insertreturning dataatexecution boolsaschar cvtnulldate alter quotes cursors cursor-movement cursor-commit cursor-name cursor-block-delete bookmark declare-fetch-commit declare-fetch-block positioned-update bulkoperations catalogfunctions bindcol lfconversion cte errors error-rollback diagnostic numeric large-object large-object-data-at-exec odbc-escapes wchar-char params-batch-exec fetch-refcursors :: --inputdir=.
connect .................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
stmthandles ................ ok   
select ..................... ok   
update ..................... ok   
commands ................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
multistmt .................. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
getresult .................. ok   
colattribute ............... ok   
result-conversions ......... ok   
prepare .................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
premature .................. ok   
params ..................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
param-conversions .......... ok   
parse ...................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
identity ................... ok   
notice ..................... ok   
arraybinding ............... ok   
insertreturning ............ ok   
dataatexecution ............ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
boolsaschar ................ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
cvtnulldate ................ ok   
alter ...................... ok   
quotes ..................... ok   
cursors .................... ok   
cursor-movement ............ ok   
cursor-commit .............. ok   
cursor-name ................ ok   
cursor-block-delete ........ ok   
bookmark ................... ok   
declare-fetch-commit ....... ok   
declare-fetch-block ........ ok   
positioned-update .......... ok   
bulkoperations ............. ok   
catalogfunctions ........... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
bindcol .................... ok   
lfconversion ............... ok   
cte ........................ ok   
errors ..................... ok   
error-rollback ............. ok   
diagnostic ................. ok   
numeric .................... ok   
large-object ............... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
large-object-data-at-exec .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
odbc-escapes ............... ok   
wchar-char ................. ok   
params-batch-exec .......... ok   
fetch-refcursors ........... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 

Test Summary Report
-------------------
connect                  (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
commands                 (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
multistmt                (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
prepare                  (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
params                   (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
parse                    (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
dataatexecution          (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
boolsaschar              (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
catalogfunctions         (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
large-object             (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
large-object-data-at-exec (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
fetch-refcursors         (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=47, Tests=47, 217 wallclock secs ( 0.10 usr  0.07 sys +  0.79 cusr  0.59 csys =  1.55 CPU)
Result: FAIL
```

**Comment by davecramer** [comment]
I guess you are running this on a mac or other *nix like machine.

The tests run fine on windows using regress.ps1

**Pull request closed without merging**